File: week3/p4.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def searchKey(head, key):
    while head:
        if head.data == key:
            return True
        head = head.next
    return False

# Example
head = Node(1)
head.next = Node(2)
head.next.next = Node(3)
head.next.next.next = Node(4)

key = 3
print("true" if searchKey(head, key) else "false")

File: week3/p4.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* newNode(int data) {
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->data = data;
    node->next = NULL;
    return node;
}

int searchKey(struct Node* head, int key) {
    while (head != NULL) {
        if (head->data == key)
            return 1; // true
        head = head->next;
    }
    return 0; // false
}

int main() {
    struct Node* head = newNode(1);
    head->next = newNode(2);
    head->next->next = newNode(3);
    head->next->next->next = newNode(4);

    int key = 3;
    if (searchKey(head, key))
        printf("true\n");
    else
        printf("false\n");

    return 0;
}
